### PR TITLE
Nhj consensus speed

### PR DESCRIPTION
**Pull request by NHJohnson**
## Description

Parallelizes operations to get mapq for consensus sites across available cores. 

Fixes # 915

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- [X] Ran on EC2 instance using files from Cavatica; expected operations seem parallelized
- [X] Consensus VCF passes visual inspection; see example output included
- Test Configuration:
    python 3.8.5
    pysam 0.16.0.1 built against samtools 1.10
    Test files:
        For add_strelka2_fields.py: https://cavatica.sbgenomics.com/u/zhangb1/kf-somatic-tools-test/files/5ffdf91de4b0a6d32f329e75/
        Sample names are: tumor: BS_F0GNWEJJ normal: BS_F5EX3ZY5
        For consensus_merge: output of add_strelka2_fields for Strelka2 VCF. Other 3 vcfs from here: https://cavatica.sbgenomics.com/u/zhangb1/kf-somatic-tools-test/files/#q?path=CONSENSUS_UPDATE_TEST&page=1&label_id=NORMED
        and this cram: https://cavatica.sbgenomics.com/u/kfdrc-harmonization/pbta-lancet-vardict-analysis/files/5d711217e4b0e4c5f8bb031c/ (must index locally)
        and hg38 reference fasta e.g. https://cavatica.sbgenomics.com/u/kfdrc-harmonization/pbta-lancet-vardict-analysis/files/5d701792e4b0950c45b0d79a/

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

File: scripts/consensus_merge.py [hunk 1]
```diff
@@ -406,21 +407,36 @@ def get_dp_consensus(dp_list):
...
             tuple of ints RMS mapq and # of mapq reads at locus
     """
+    if cram_path.endswith('cram'):
+        if not reference:
+            raise IOError('Must provide reference with CRAM')
+        cram = pysam.AlignmentFile(cram_path, 'rc', reference_filename=reference)
+    elif cram_path.endswith('bam'):
+        cram = pysam.AlignmentFile(cram_path, 'rb')
+    else:
+        raise IOError('File provided to "cram" argument must have .cram or .bam extension')
```
Comment (dmiller15) [inline scripts/consensus_merge.py:434]: So would the old method of giving this function a `pysam.AlignmentFile` not work/work slower with the multiprocessing?

**Review by dmiller15 — APPROVED** [review]
Looks good

File: scripts/consensus_merge.py (hunk 1 above)
Comment (NHJohnson) [reply scripts/consensus_merge.py:434]: With `multiprocessing` you can only pass around objects that can be serialized with `pickle`. This is often an annoying limitation as random objects like AlignmentFiles turn out not to pickle. And it turns out there's no profound reason for it! [Some guy](https://pypi.org/user/mmckerns/) produced a module called `dill` that _can_ serialize arbitrary python objects, and he made a fork of `multiprocessing` on top of that overcomes these limitations, which I have used before. It's not exactly a standard-library module, though, and I decided not to use it here on the theory that there wouldn't be any major speed difference, but I didn't actually try to find out.

**Review by sickler-alex — COMMENTED** [review]
While you're updating this, it might be worth taking the time to cleanup the code a little. Running `pylint` on the script shows a bunch of trailing whitespace, bad indentations, and unused variables.

File: scripts/consensus_merge.py (hunk 1 above)
Comment (dmiller15) [reply scripts/consensus_merge.py:434]: Cool! Thanks for the lesson. I've never used multiprocessing before so this is good knowledge to have.

**Review by migbro — COMMENTED** [review]
Alluding to @sickler-alex pylint comment, it may be a hint to something missing, like here: `output_record.samples[index]['ADR'] = (ad_range_1, ad_range_1)`
I am guessing should be `output_record.samples[index]['ADR'] = (ad_range_1, ad_range_2)`. Running a local test now for performance :)

**Comment by NHJohnson** [comment]
Good catch on the 'ADR' thing. I got my pylint score up to 7.03 from 5.61! I'm going to leave it there, though, it made some valid points but also has very specific ideas about what variables should be called.

**Review by sickler-alex — APPROVED** [review]

**Comment by NHJohnson** [comment]
@migbro How many cores do you have?

**Comment by NHJohnson** [comment]
After some more poking around and trying things out, I've concluded it's not possible to serialize the pysam AligmentFile object, and so having each process open and close the file is the only way to go. I did empirically observe that supplying multiple_iterators=True to fetch() does not appear to be necessary, and this results in a small speedup. The whole process is not as fast as I would like, but the situation is rescued somewhat by a few considerations:

- @migbro was able to learn that the inputs we have been testing on are much larger than is typical for this workflow, so the time required in practice will be much shorter than what we have been experiencing
- In CWL we can potentially detect larger-than-normal inputs and surge to an instance with more cores to take better advantage of the parallelization in those cases where it would make a real difference. 

For these reasons I will merge this today unless anyone would like to discuss further.

**Pull request merged**